spring-integration-r2dbc/src/main/java/org/springframework/integration/r2dbc/inbound/R2dbcMessageSource.java

```diff
@@ -151,9 +175,24 @@ protected Object doReceive() {
 				Mono.fromSupplier(() -> this.queryExpression.getValue(this.evaluationContext))
 						.map(this::prepareFetch);
 		if (this.expectSingleResult) {
-			return queryMono.flatMap(FetchSpec::one);
+			return queryMono.flatMap(FetchSpec::one)
+					.flatMap(this::executeUpdate);
+		}
+
+		return queryMono.flatMapMany(FetchSpec::all)
+				.flatMap(this::executeUpdate);
+	}
+
+	private Mono<Object> executeUpdate(Object result) {
+		if (this.updateSql != null) {
```

Comment:
Since this option is static and doesn't depend on a value we should apply `if...else` logic only once.
See `Flux.transform()` operator:

```
	 * Transform this {@link Flux} in order to generate a target {@link Flux}. Unlike {@link #transformDeferred(Function)}, the
	 * provided function is executed as part of assembly.
```